### PR TITLE
feature: add docker and database setup and fix interface

### PR DESCRIPTION
**Pull request by kleyow**
Implementing the CRUD handlers along with docker and the database setup was getting cumbersome.
So I'm spliting it.
This PR covers 
- docker and docker-compose setup
- table schema (please look over `data/database/testing.sql` for tables. I have an irking feeling that these tables aren't sufficient maybe)
- config revamp and database config
- db helper class & error (pulled from als-oracle-pathfinder)
- aligned the server initialization more to the als-oracle-pathfinder repo, might need revisiting
- fixing the oracle interface schema
- re-write tests in jest

I so need to revisit the template...I'm make a ticket for that...

### CODE REVIEW

File: docker/wait4/wait4.config.js [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+  // services definitions
+  services: [
+    {
+      name: 'als-consent-oracle',
+      wait4: [
+        {
+          description: 'MySQL Consent Oracle',
+          uri: 'oracle-mysql:3306',
+          method: 'mysql',
+          // customized RC setup
```
Comment (eoln) [inline docker/wait4/wait4.config.js:21]: if we drop RC config there will be a need to refactor RC based wait4 methodMySQL strategy for knex or use `ncat` strategy

File: docker/wait4/wait4.js [hunk 2]
```diff
@@ -0,0 +1,179 @@
...
+      })
+    }
+    // no more retries left
+    console.error(`wait4 Out of retries for uri:${waitJob.uri}\n\t\tand method: ${waitJob.method}`)
+    waitJob.status = 'stalled'
+    return Promise.reject(err)
+  }
+}
+
+/**
```
Comment (eoln) [inline docker/wait4/wait4.js:106]: if we decide to use Convict there is a need to refactor this function

File: package.json [hunk 3]
```diff
@@ -21,7 +21,9 @@
         "@mojaloop/central-services-logger": "^10.6.0",
```
Comment (eoln) [inline package.json:21]: there is a need to fix `name`, `author`, `repository` properties!

File: src/index.js [hunk 4]
```diff
@@ -1,42 +1,40 @@
...
+const util = require('util')
+const Logger = require('@mojaloop/central-services-logger')
+const Config = require('./config')
+const { createServer } = require('./server')
+const { OracleDatabase } = require('./db')
 
-const Program = new Command()
+const initialize = async () => {
+  // Set log level to info
+  // eslint-disable-next-line
```
Comment (eoln) [inline src/index.js:11]: is this linting disable needed?

File: src/index.js [hunk 5]
```diff
@@ -1,42 +1,40 @@
...
+  const oracleDb = new OracleDatabase(Config.DATABASE)
+  const server = await createServer(Config.SERVER_PORT, oracleDb)
 
-Program.command('api')
-  .alias('a')
-  .description('Start the Transaction Requests API. Use options to specify server type of none to run both') // command description
+  if (server) {
+    try {
+      // Perform initial health check
+      const unhealthy = await server.app.healthCheck()
```
Comment (eoln) [inline src/index.js:20]: this reversed logic makes code foggy

File: src/index.js [hunk 6]
```diff
@@ -1,42 +1,40 @@
...
+}
 
-if (Array.isArray(argv) && argv.length > 1) {
-  // parse command line vars
-  Program.parse(argv)
-} else {
-  // display default help
-  Program.help()
+try {
+  initialize()
```
Comment (eoln) [inline src/index.js:37]: initialize is an async method, if don't want to wait on a promise to resolve, let leave a comment here why...

File: test/unit/plugins.test.js [hunk 7]
```diff
@@ -8,33 +8,24 @@
...
-let sandbox
 describe('plugins', () => {
-  beforeAll(() => {
-    sandbox = Sinon.createSandbox()
-  })
-
-  afterEach(() => {
-    sandbox.restore()
+  beforeEach(() => {
+    jest.clearAllMocks()
```
Comment (eoln) [inline test/unit/plugins.test.js:18]: maybe let setup in `jest.config` automatic clearing of mocks... https://jestjs.io/docs/en/configuration#clearmocks-boolean

File: test/unit/server.test.js [hunk 8]
```diff
@@ -8,7 +8,29 @@
  *       Lewis Daly - lewisd@crosslaketech.com                            *
  **************************************************************************/
 'use strict'
+const { OracleDatabase } = require('../../src/db')
+const { createServer } = require('../../src/server')
 
+const fakeConfig = {
```
Comment (eoln) [inline test/unit/server.test.js:14]: I would use here real config, elsewhere there is a need to recheck this fakeConfig for every change in the real one....

**Comment by eoln** [comment]
why we dropped using Typescript to implement new services?
I would borrow as much code as possible from the auth-service...

**Review by eoln — DISMISSED** [review]

File: Dockerfile [hunk 9]
```diff
@@ -0,0 +1,41 @@
...
+
+WORKDIR /opt/als-consent-oracle
+
+RUN apk add --no-cache -t build-dependencies git make gcc g++ python libtool autoconf automake bash mysql-client \
+    && cd $(npm root -g)/npm \
+    && npm config set unsafe-perm true \
+    && npm install -g node-gyp
+
+COPY ./package.json ./package-lock.json /opt/als-consent-oracle/
+RUN npm install --production
```
[the remark this answers is not in the document]
Comment (kleyow) [reply Dockerfile]: done

File: Dockerfile [hunk 10]
```diff
@@ -0,0 +1,41 @@
...
+LABEL org.label-schema.vcs-ref=$VCS_REF
+LABEL org.label-schema.url="https://mojaloop.io/"
+LABEL org.label-schema.version=$VERSION
+
+# Create a non-root user: als-oracle-user
+RUN adduser -D als-oracle-user
+USER als-oracle-user
+
+COPY --chown=als-oracle-user --from=builder /opt/als-consent-oracle .
+RUN npm prune --production
```
[the remark this answers is not in the document]
Comment (kleyow) [reply Dockerfile]: done

File: src/config.js [hunk 11]
```diff
@@ -0,0 +1,65 @@
...
+ *  (C) Copyright Mojaloop Foundation 2020 - All rights reserved.         *
+ *                                                                        *
+ *  This file is made available under the terms of the license agreement  *
+ *  specified in the corresponding source code repository.                *
+ *                                                                        *
+ *  Original Author:                                                      *
+ *       Kevin Leyow - kevin.leyow@modusbox.com                           *
+ **************************************************************************/
+'use strict'
+const RC = require('parse-strings-in-object')(require('rc')('ALS', require('../config/default.json')))
```
[the remark this answers is not in the document]
Comment (kleyow) [reply src/config.js:11]: Isn't `rc` mojaloop's standard atm?

File: docker-compose.yml [hunk 12]
```diff
@@ -0,0 +1,47 @@
+version: "3.7"
+
+networks:
+  oracle-mojaloop-net:
```
[the remark this answers is not in the document]
Comment (kleyow) [reply docker-compose.yml]: done.

File: package.json (hunk 3 above)
Comment (kleyow) [reply package.json:21]: done

File: src/index.js (hunk 4 above)
Comment (kleyow) [reply src/index.js:11]: for now yes to ignore an "Arrow function should not return assignment". this was code I took from the pathfinder repo, so I didn't want to change it up here.

File: src/index.js (hunk 5 above)
Comment (kleyow) [reply src/index.js]: reversed. done

File: src/index.js (hunk 6 above)
Comment (kleyow) [reply src/index.js:37]: for now yes to ignore an "Arrow function should not return assignment". this was code I took from the pathfinder repo, so I didn't want to change it up here.

File: test/unit/plugins.test.js (hunk 7 above)
Comment (kleyow) [reply test/unit/plugins.test.js]: done

File: test/unit/server.test.js (hunk 8 above)
Comment (kleyow) [reply test/unit/server.test.js]: done

**Comment by kleyow** [comment]
> why we dropped using Typescript to implement new services?
I would borrow as much code as possible from the auth-service...

I brought typescript up when discussing the ticket, I believe Lewis said something along the lines that  the oracle didn't warrant the time needed vs just getting the oracle operational.

I'm all for it but the day estimation was low for this epic.

I can come back around and convert it in my idle time. Or convert it after this ticket. Whatever is agreed upon.

File: src/config.js (hunk 11 above)
[the remark this answers is not in the document]
Comment (lewisdaly) [reply src/config.js:11]: It's probably something we should bring up with the DA, but we seem to use RC sometimes, custom parsers other times, and lately Convict on the pisp stuff.

File: src/config.js [hunk 13]
```diff
@@ -0,0 +1,65 @@
...
+
+      // free resouces are destroyed after this many milliseconds
+      idleTimeoutMillis: getOrDefault(RC.DATABASE.IDLE_TIMEOUT_MILLIS, 30000),
+
+      // how often to check for idle resources to destroy
+      reapIntervalMillis: getOrDefault(RC.DATABASE.REAP_INTERVAL_MILLIS, 1000),
+
+      // long long to idle after failed create before trying again
+      createRetryIntervalMillis: getOrDefault(RC.DATABASE.CREATE_RETRY_INTERVAL_MILLIS, 20)
+      // ping: function (conn, cb) { conn.query('SELECT 1', cb) }
```
Comment (lewisdaly) [inline src/config.js:58]: nit: remove commented code

File: src/db.js [hunk 14]
```diff
@@ -0,0 +1,72 @@
...
+      const result = await this.client.raw('SELECT 1 + 1 AS result')
+      return !!(result)
+    } catch (err) {
+      return false
+    }
+  }
+}
+
+/* class OracleDbError extends Error {
+  constructor (code, message) {
```
Comment (lewisdaly) [inline src/db.js:30]: nit: remove commented lines

File: src/db.js [hunk 15]
```diff
@@ -0,0 +1,72 @@
...
+    super(message)
+    this.code = code
+  }
+}
+*/
+const ErrorCodes = Object.freeze({
+  PARTICIPANT_NOT_FOUND: 'PARTICIPANT_NOT_FOUND'
+})
+/*
+const Errors = {
```
Comment (lewisdaly) [inline src/db.js:40]: nit: remove commented lines

File: src/error.js [hunk 16]
```diff
@@ -0,0 +1,15 @@
+
+'use strict'
+
+const Boom = require('@hapi/boom')
+
+const responses = {
+  // Investigate why central-services-error-handling turns 404's into method not allowed errors
```
Comment (lewisdaly) [inline src/error.js:7]: Is this still a valid comment?

**Comment by kleyow** [comment]
I've actually been working on the typescript upgrade and just taking files from `auth-service` which was pretty easy to pull over.

Closing this in favour of https://github.com/mojaloop/als-consent-oracle/pull/2 (still working on it though)

So this sort of outdated now :x we'll resume the convo when the PR is ready.

**Comment by lewisdaly** [comment]
Ok, thanks @kleyow ... get some sleep!

**Pull request closed without merging**